go/parquet/file/column_writer_test.go

```diff
@@ -45,6 +49,73 @@ const (
 	DictionaryPageSize = 1024 * 1024
 )
 
+type mockpagewriter struct {
+	mock.Mock
+}
+
+func (m *mockpagewriter) Close(hasDict, fallBack bool) error {
+	return m.Called(hasDict, fallBack).Error(0)
+}
+func (m *mockpagewriter) WriteDataPage(page file.DataPage) (int64, error) {
+	args := m.Called(page)
+	return int64(args.Int(0)), args.Error(1)
+}
+func (m *mockpagewriter) WriteDictionaryPage(page *file.DictionaryPage) (int64, error) {
+	args := m.Called(page)
+	return int64(args.Int(0)), args.Error(1)
+}
+func (m *mockpagewriter) HasCompressor() bool {
+	return m.Called().Bool(0)
+}
+func (m *mockpagewriter) Compress(buf *bytes.Buffer, src []byte) []byte {
+	return m.Called(buf, src).Get(0).([]byte)
+}
+func (m *mockpagewriter) Reset(sink utils.WriterTell, codec compress.Compression, compressionLevel int, metadata *metadata.ColumnChunkMetaDataBuilder, rgOrdinal, columnOrdinal int16, metaEncryptor, dataEncryptor encryption.Encryptor) error {
+	return m.Called().Error(0)
+}
+
+func TestWriteDataPageV2NumRows(t *testing.T) {
```

Comment:
this seems OK, one other note of something broken in C++ for V2 is rows are supposed to begin and end on page boundaries (V1 doesn't have this gaurantee)

Reply:
that's an interesting one that makes the logic for `WriteBatch` a bit more complex and annoying. I guess that's probably why it's broken for C++ currently haha. I'll see if i can figure out a way to slightly modify the logic for the batching so that I always do batches on row boundaries for V2. if not then i'll just document the limitation, if it's currently broken in the C++ too then I don't feel bad about kicking that can down the road for now in the interest of getting this merged. I'll comment back with what I end up doing.

Reply:
Ok cool, I solved it by mucking with the batching logic to ensure that the batches in `WriteBatch` always begin at a row boundary (writing slightly smaller batches if necessary). this lets everything else stay the same since as long as I guarantee that the batch I write starts with a new row, and we only check to write the page after writing a batch, it guarantees that we never have a row crossing page boundaries.

the only drawback is that doing it this way means that if you are writing a repeated column with DataPageV2, `WriteBatch` must always start with a new row. that keeps the guarantee. I'm ok with that though and I've updated the documentation accordingly and added a test to ensure that batching makes sure that rows don't cross page boundaries.

Assuming nothing goes wrong with the tests in the CI, i'll merge this after everything passes.